### DISCUSSION TITLE
Supported media types for incoming messages

### DISCUSSION THREAD

**Discussion by dominikklein** [ignacio-chiazzo/ruby_whatsapp_sdk#127] [General]
I know that we have this list of supported media types, which are also correctly implemented: https://developers.facebook.com/docs/whatsapp/cloud-api/reference/media

When I saw it correctly the types are checked for incoming messages inside the gem here, but for me, it looks like this restriction is only for outgoing messages, because here also the API returns error messages for media types outside of this defined list.

I see that for incoming messages the payloads have all the correct data for other media types and manually I can also get the media information from the media id and download this kind of file with the given media url.

What are you thinking about this situation?

**Comment by ignacio-chiazzo** [comment]
Hi,

could you elaborate more on what incoming means here? Do you mean that  to upload a media you must submit a supported type, but when querying a media you could get an unsupported type?

**Comment by dominikklein** [comment]
I mean when we receiving a new message via webhook. Then next step is to use the media-id to get the meta information of the attachment/media and then next step to download the file.
Inside the incoming media message via the webhook, we are also receiving messages which not restricted related to this media list inside the documentation.
But it's checked before the download in: https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/fbe074fabdc15e334f4ec3395c1a8b11bfcfe970/lib/whatsapp_sdk/api/medias.rb#L71

So for my understanding this defined media types in the whatsapp cloud api looks for me only for restriction in the outgoing part (upload the media).

**Comment by ignacio-chiazzo** [comment]
Interesting. I haven't seen any examples for downloading other media types. If that's possible then we'd need to remove the restriction for download. 

Did you manage to reproduce it?

**Comment by dominikklein** [comment]
Yes, I tested a lot of different file types, for example this is a incoming message on the webhook with some special type:

```
{
    "object": "whatsapp_business_account",
    "entry": [
        {
            "id": "260895230431646",
            "changes": [
                {
                    "value": {
                        "messaging_product": "whatsapp",
                        "metadata": {
                            "display_phone_number": "???",
                            "phone_number_id": "???"
                        },
                        "contacts": [
                            {
                                "profile": {
                                    "name": "John Doe"
                                },
                                "wa_id": "???"
                            }
                        ],
                        "messages": [
                            {
                                "from": "???",
                                "id": "wamid.HBgNNDkxNTEyNTIyNDgwMRUCABIYFjNFQjAyQ0U3MEVDRjhFMzlFRDk1NzkA",
                                "timestamp": "1708678909",
                                "type": "document",
                                "document": {
                                    "filename": "Questions.drawio",
                                    "mime_type": "application/octet-stream",
                                    "sha256": "rg4rWJDoZdXDOQGw5MAmJBmpESyw9klyUyGI7LD5tdA=",
                                    "id": "727074052551625"
                                }
                            }
                        ]
                    },
                    "field": "messages"
                }
            ]
        }
    ]
}
```

**Reply by ignacio-chiazzo** [reply]
Ohhhh! How come was `octet-stream` created 🤔. That seems worth raising an issue on the Meta issues tracker.

**Reply by ignacio-chiazzo** [reply]
If this is the case, you're right we should not validate any incoming media type.

**Reply by dominikklein** [reply]
I tested also with zip or other types :-D

```
...
"document": {
    "caption": "123123",
    "filename": "area-code.rb",
    "mime_type": "text/x-ruby-script",
    "sha256": "/2fOXXlwTgxM2XIY/KT/nqKZm6urmJPy+p6gCRW08K4=",
    "id": "1090325068775431"
}
...
```

```
...
"document": {
    "filename": "contentful-plugin.zip",
    "mime_type": "application/zip",
    "sha256": "FxDRCHtAmCaImNDrYlIv3V7GWsc0O3iew1uSAbVQH+I=",
    "id": "724391469806431"
}
...
```

**Reply by ignacio-chiazzo** [reply]
and how did you send the unsupported media type? via a phone in whatsapp?

**Reply by dominikklein** [reply]
I tested with my phone (iPhone) app and also with the Mac WhatsApp client and with the web client.

**Comment by ignacio-chiazzo** [comment]
Thanks for reporting @dominikklein. We should remove the validation for now when downloading the image. 

If you want to contribute, all we need to do is: 
1) comment the [line here ](https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/main/lib/whatsapp_sdk/api/medias.rb#L72). 
2) Skip the tests that assert raising an error when the media is invalid. 
3) Add a comment t[o the `Supported Media File` ](https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/fbe074fabdc15e334f4ec3395c1a8b11bfcfe970/lib/whatsapp_sdk/resource/media_types.rb#L27).

**Reply by dominikklein** [reply]
Yes, we will try to add a PR today :-)

**Reply by fliebe92** [reply]
https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/pull/128 🎉

**Reply by ignacio-chiazzo** [reply]
Thank you both!
